### ISSUE TITLE
[20] Large Language Model based Long-tail Query Rewriting in Taobao Search

### ISSUE THREAD

**Issue by noparkee** [noparkee/paper#16]
<img width="700" alt="스크린샷 2024-12-24 오후 3 15 14" src="https://github.com/user-attachments/assets/e85bb62b-9be5-4632-9c77-81520360fc73" />

[paper](https://arxiv.org/abs/2311.03758)

---

- e-commerce 검색에서 롱테일 쿼리 분석 및 이와 관련있는 semantic gap 문제 파악. 산업용 query rewriting 을 위해서 llm을 처음으로 파인튜닝
- 롱테일 쿼리의 semantic gap 문제를 해결하기 위해 three-stage fine-tuned framedwork (BEQUE) 제시
- 오프라인 및 온라인 실험을 통해 입증. 전자상거래 수익을 크게 향상시킬 수 있는 능력

---

<img width="700" alt="스크린샷 2024-12-24 오후 3 32 40" src="https://github.com/user-attachments/assets/8bbcc389-ec46-4c3d-9c78-79cd4674d1fe" />

#### Query Rewriting Dataset
<img width="469" alt="스크린샷 2024-12-24 오후 3 43 02" src="https://github.com/user-attachments/assets/a85eb5f8-3929-446b-963f-65c38545f096" />

- 유저가 x를 입력하면 타오바오의 old rewriting policy가 rewrites queries Y = {y1, y2, ..., yn} 를 작성
- 이 리스트에서 최상위 순위 y1을 gold standart candidate로 선정하여 N개의 샘플로 initial rewriting dataset 생성

<img width="428" alt="스크린샷 2024-12-24 오후 3 43 09" src="https://github.com/user-attachments/assets/9cd31e21-81f7-485a-9f5f-689660a1340a" />

- 쿼리와 rewrite 사이의 semantic similarity가 비슷한 상품의 검색을 보장하지는 않음
- rewrite y에서 검색된 상품과 원래 쿼리 x 사이의 높은 연관성 (즉, 질의 단위의 유사성이 아니라 검색 결과와 원질의의 유사성)
- rejection sampling 사용
- $rele$과 $\tau^{rele}$는 relevance method와 그 threshold를 의미

<img width="352" alt="스크린샷 2024-12-24 오후 3 43 18" src="https://github.com/user-attachments/assets/3d82f933-ae15-44f6-b31c-1d5c423b4cff" />

- 타오바오의 previous-generation of rewriting models는 롱테일 쿼리에 대한 최적화가 부족했음.
- rejection sampling을 다시 활용해서 $\mathcal{D}_r$를 필터링
- 또한 롱테일 쿼리를 더 잘 처리하기 위해서 가장 최근에 상호작용한 상품의 title을 supplementary information으로 포함시킴
- $\mathcal{E}_x$는 쿼리 x가 상호 작용한 상품 리스트, $incr$, $\tau^{incr}$는 각 query-rewrite pair의 increment 방법과 threshold

#### Auxiliary Task Datasets
<img width="401" alt="스크린샷 2024-12-24 오후 5 55 10" src="https://github.com/user-attachments/assets/d21fd5dd-7b52-4fd5-8359-f1f92245f228" />

- 롱테일 쿼리를 이해하는 LLM 능력을 향상 시키기 위해 quality classification, product title prediction, Chain-of-thought 데이터셋을 모음.
  - quality classfication: online logs에서 query pair를 추출 -> query pair에 사람이 주석을 달아 sft에 지정된 데이터 요구 사항을 충족하는지 확인
  - product title prediction: 쿼리가 가장 최근에 상호작용한 상품을 선택해서 <query, product title> pair 생성
  - CoT task: 사람 평가자를 위한 prompt를 구축하기 위해서 원본 온라인 쿼리를 사용. 이 평가자들은 쿼리 검색 품질을 개선하기 위해 query rewrites를 제시할 뿐만 아니라 구체적인 근거를 설명하고, 자신의 사고 과정을 명확하게 표현해야함.

#### Supervised Fine Tuning

<img width="370" alt="스크린샷 2024-12-24 오후 5 56 11" src="https://github.com/user-attachments/assets/5e396982-4ee9-4709-80bf-6e5c3259c24f" />

- 프롬프트 x와 gold standard y가 주어지고, p(y|x)를 최대화 하는 것이 목표
- $\mathcal{D}_{msft}$는 query rewriting dataset과 다양한 auxiliary task dataset으로 구성된 데이터셋을 의미하고, $\pi$와 $\theta$는 query rewriting 모델의 파라미터를 의미
- 프롬프트 x에 대응하는 loss는 무시

### Offline Feedback

- 대부분의 alignment method는 manual annotation과 training-based reward model에 의존하고 있음. 하지만, 이런 접근 방식은 annotation의 품질과 reward model tratining에 영향을 받을 수 있고 이로 인해 response score를 부정확하게 반영하고 생성 모델의 학습을 손상시킬 수도 있음.
- rewrite의 퀄리티를 측정하기 위해 relevance, increment, hitrate 라는 세 가지 점수 제안
  - relevance: 모델이 사용자의 원래 의도와 완전히 다른 제품을 검색하는 것을 방지하기 위함
  <img width="271" alt="스크린샷 2024-12-24 오후 6 11 26" src="https://github.com/user-attachments/assets/09369c2d-4f41-4e39-ba55-1dadf4f60f3a" />

    - $f(-)$는 상품 타이틀과 query text 사이의 관련성을 평가하기 위한 관련성 함수, $\tau'$는 threshold, $\mathcal{Z}_y$는 텍스트 y의 오프라인 검색 상품 목록

  - increment
  <img width="461" alt="스크린샷 2024-12-24 오후 6 11 50" src="https://github.com/user-attachments/assets/5e0972d9-16b8-4365-ba43-fc7365e41f56" />

    - $\mathcal{Z}_x$는 텍스트 x에 대한 오프라인 검색 상품 목록, $\mathcal{Z}_e$는 타오바오 검색 그룹에서 관리하는 excellent 상품 셋

  - hitrate: rewrite가 원래 쿼리의 semantic gap을 어느 정도 보완할 수 있는지
  <img width="301" alt="스크린샷 2024-12-24 오후 6 11 58" src="https://github.com/user-attachments/assets/eb35d9d0-eb36-4a05-aab8-79ac1b8130e6" />

    - $\mathcal{E}$는 사용자가 검색 시나리오 외에 거래한 제품 모음, $\mathcal{Z}_x$와 $\mathcal{Z}_y$는 텍스트 x와 y의 오프라인 검색 제품 셋을 의미
    - 현재 사용자의 쿼리와 의미적으로 관련있는 제품이 검색에서 거래되지 않았다고 가정. 이는 원래 쿼리가 제품을 검색하지 못했음을 의미. 하지만 rewrite가 제품을 return 할 수 있다면 제품이 $\mathcal{Z}_x \cup \mathcal{Z}_y$에 존재한다는 의미

### Objective Alignment

- reward model에 의한 bias를 방지하기 위해 브래들리 테리 모델에 기반한 Preference Rank Optimization (PRO) 도입
- 오프랑니 피드백에 의해 rewrite partial order을 학습하는 것을 목표.
- 브래들리 테리 모델에 다르면, policy를 선택할 확률은 reward에 비례해야함.
- y1 > y2가 주어지면 선호 확률은 다음과 같다.
   <img width="244" alt="스크린샷 2024-12-26 오후 3 29 32" src="https://github.com/user-attachments/assets/354b25a0-41f3-457e-92f8-169e4cdafbc3" />

  - $r$은 보상 함수로, PRO에서 LLM에 의해 생성된 rewrite의 normalized log probability

- PRO는 이 pairwise partial order를 listwise partial order로 확장.
- reward 기반의 ranking의 중요도를 파악하기 위해 temperature 도입
  <img width="386" alt="스크린샷 2024-12-26 오후 4 03 09" src="https://github.com/user-attachments/assets/a77d000d-4b35-4282-907f-7d939cc57e28" />
  <img width="302" alt="스크린샷 2024-12-26 오후 4 53 15" src="https://github.com/user-attachments/assets/91777e1b-2309-4125-8853-31c40fddc244" />
    - $\tau_k^i$와 $\tau_k^k$는 ranking difference, $\mathcal{D}_{PRO}$는 alignment를 위한 데이터셋, $\pi_{PRO}$와 $\theta$는 policy model과 파라미터, $n$는 candidate rewrites 개수

- 모델이 정상적인 output을 생성할 수 있는 능력을 유지하기 위해 SFT loss는 PRO loss가 더해짐
<img width="182" alt="스크린샷 2024-12-26 오후 4 52 54" src="https://github.com/user-attachments/assets/e1334fd1-2a8b-4b00-9f70-be534733472f" />

### Online Serving

- torso, tail 질의에 대해서 offline BEQUE 활용하여 query rewrite 수행
- 이 쿼리에서 생성된 rewrites는 key-value로 저장되어 online으로 서빙할 수 있음

---

#### Training Dataset
- SFT training
  - 타오바오 온라인 rewrite log에서 약 2천만개의 record를 처음 SFT 훈련 데이터로 추출
  - 학습 데이터의 품질을 높이기 위해 두 번의 rejection sampling -> 419,806 쌍의 <query, rewrite> 데이터셋
  - 155,662개의 manually rewriting data 포함시켜 SFT 모델의 rewrite가 사람의 선호도와 align 되는지 확인
  - quality classification, title prediction, CoT task에서 각 5만개의 샘플을 query rewrite 데이터와 결합하여 데이터셋 구축
- objective alignment 
  - query rewrite 데이터셋에서 만개의 쿼리를 랜덤하게 선택하여 후보 생성 (각 쿼리에 대해 SFT 모델을 사용하여 5개의 후보 생성)
  - 타오바오 오프라인 시스템을 통해 5만개의 rewrite query에 score
  - 이상치 제거 후 45,350개의 candidate rewrite 훈련데이터

#### Test Dataset
- 오프라인 테스트를 위해서 타오바오 검색 로그에서 14,981개의 쿼리
  - 절반은 실제 쿼리에 비례하여 랜덤 샘플, 나머지 절반은 롱테일 쿼리에서 샘플링

### Offline Experiments
####Ablation Study of Multi-instruction SFT data
- 로그에서 pair를 뽑아서 이것으로만 학습했을 때 (rejection sampling도 적용하지 않고, 위의 다른 태스크 데이터셋들과 혼합하지 않음) 성능이 더 낮음 => rejection sampling을 통해 query pair의 품질 향상, 다른 태스크들을 통해 모델의 이해를 강화하여 더 나은 품질의 rewrite

#### Results of Different Contrast Number
- candidates 가 증가할 수록, relevance는 상승하지만 increment와 hitrate는 감소하는데 이는 모든 candidates를 gold standard로 간주하기 때문이라 볼 수 있음 => candidates 풀이 클수록 SFT loss weight는 증가하고 partial order loss 가중치는 감소하여 생성된 rewrite 의 관련성 향상

#### Main Results
<img width="808" alt="스크린샷 2024-12-26 오후 7 24 43" src="https://github.com/user-attachments/assets/61909b31-2056-426d-8957-4978776d3223" />

- Generative models outperform discriminative models when rewriting “Torso” and “Tail” queries.
  - discriminative model은 기존 쿼리의 rewrite 역시 로그 상에 존재하기 때문에 torso나 tail에서는 그런 의미적으로 유사한 rewrite이 없는 경우가 많다. 하지만 generative model은 로그 외의 질의를 생성할 수 있음
- LLMs exhibit superior long-tail semantic understanding capabilities compared to small models.
  - BART와 Qwen 비교
- Retrieval augmentation methods demonstrate limited semantic expansion capabilities.
  - Q2D와 BEQUE 비교
- The reinforcement learning (RL) may introduce bias and impact the effectiveness of the rewriting LLMs.
  - 흠 (?)
  - RL은 base model의 학습을 가이드하기 위해 reward model을 도입하는데, reward를 계산하기 위해서는 오프라인 검색 시스템 시뮬레이션 필요하고 reward model이 검색 시스템의 특징을 정확히 포착하지 못 할 수 있음. -> RL 모델 성능 저하 될 수도
  - BEQUE는 contrastive learning을 통해 후보의 partial order를 명시적으로 학습하여 reward model로 인한 bias를 피할 수 있음.
- Different offline metrics work differently as rewards.

### Online Experiments
<img width="430" alt="스크린샷 2024-12-26 오후 7 26 56" src="https://github.com/user-attachments/assets/26a0e883-4697-4c5c-b0ab-d2eb9561594f" />

- 오프라인에서 inference 하기 때문에 rewrite 못하는 online query가 약 70%
- 약 27%에 해당하는 query가 BEQUE에 의해 처리되고 이 경우 GMV, #Trans, UV가 각각 2.96%, 1.36%, 1.22% 증가

---

### Conclusion
- BEQUE의 주요 목표는 semantic matching process 중에 발생하는 semantic gap을 해결하는 것 (특히 롱테일 쿼리)
- rejection sampling과 auxiliary task mixing을 통해 rewrite 데이터 셋의 품질 개선 -> 정제된 데이터셋을 이용하여 LLM 학습 -> 학습된 LLM을 이용해서 각 샘플된 쿼리마다 candidates 생성 -> candidates 사이의 partial order를 만들기 위해 타오바오 서치 기반의 offline feedback 시스템 생성 -> 이 피드백 시스템을 이용해서 relevance, increment, hitrate 관점에서 평가 -> 이 partial order를 이용해서 모델의 목표를 타오바오의 목표와 align되는 PRO 도입
- 위 과정을 통해 high-quality retrieval outcomes를 만들어내는 rewrite 생성

**Comment by noparkee** [comment]
- auxiliary task mixing을 통해 llm이 e-commerce, query에 대한 이해도를 높임.
- partial order를 만들기 위한 offline feedback 시스템
  - beam search를 통해서 candidates를 만든다고 되어 있는데 이렇게 만들어진 candidates를 offline feedback을 통해 relevance, increment, hitrate 점수를 얻고 이를 적당히 가중합해서 partial order를 매기는건가?
